mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -0,0 +1,754 @@
+            };
+
+            Ok(TxAmounts {
+                total: amount,
+                my_balance_change: received_by_me.clone() - spent_by_me.clone(),
+                spent_by_me,
+                received_by_me,
+            })
+        }
+
+        fn get_fee_details<Coin>(fee: Fee, coin: &Coin) -> Result<TxFeeDetails, String>
+        where
+            Coin: TendermintTxHistoryOps,
+        {
+            let fee_coin = fee
+                .amount
+                .first()
+                .ok_or_else(|| "fee coin can't be empty".to_string())?;
+            let fee_amount: u64 = fee_coin.amount.to_string().parse().map_err(|e| format!("{:?}", e))?;
+
+            Ok(TxFeeDetails::Tendermint(TendermintFeeDetails {
+                coin: coin.platform_ticker().to_string(),
+                amount: big_decimal_from_sat_unsigned(fee_amount, coin.decimals()),
+                gas_limit: fee.gas_limit.value(),
+            }))
+        }
+
+        fn upsert_tx_details(
+            details_map: &mut HashMap<String, Vec<TransactionDetails>>,
+            new_details: TransactionDetails,
+        ) {
+            match details_map.entry(new_details.coin.clone()) {
+                Entry::Vacant(e) => {
+                    e.insert(vec![new_details]);
+                },
+                Entry::Occupied(mut e) => {
+                    e.get_mut().push(new_details);
+                },
+            };
+        }
+
+        struct ParseTxDetailsArgs<Coin, Storage, 'a> {
+            coin: &'a Coin,
+            storage: &'a Storage,
+            active_assets: &'a HashMap<String, String>,
+            internal_id: BytesJson,
+            block_height: u64,
+            tx_hash: String,
+            tx_body_msg: &'a [u8],
+            fee_details: TxFeeDetails,
+        }
+
+        async fn parse_send_tx_details<Coin, Storage>(
+            args: ParseTxDetailsArgs<Coin, Storage, '_>,
+        ) -> Result<Option<TransactionDetails>, StopReason>
+        where
+            Coin: TendermintTxHistoryOps,
+            Storage: TxHistoryStorage,
+        {
+            let sent_tx = MsgSend::decode(args.tx_body_msg).map_err(|e| StopReason::Marshaling(e.to_string()))?;
+            let coin_amount = sent_tx
+                .amount
+                .first()
+                .ok_or_else(|| StopReason::Marshaling("amount can't be empty".to_string()))?;
+            let denom = coin_amount.denom.to_lowercase();
+
+            let tx_coin_ticker = match args.active_assets.get(&denom) {
+                Some(t) => t,
+                None => return Ok(None),
+            };
+
+            let wallet_id = WalletId::new(tx_coin_ticker.replace('-', "_"));
+            if matches!(
+                args.storage.get_tx_from_history(&wallet_id, &args.internal_id).await,
+                Ok(Some(_))
+            ) {
+                log::debug!("Tx '{}' already exists in tx_history. Skipping it.", &args.tx_hash);
+                return Ok(None);
```

Comment:
It's quite unclear that the function returns `Ok(None))` if the transaction is in the storage already.
Also this check is duplicated at the several places.
What do you think about moving this check outside of `parse_send_tx_details`, `parse_create_htlc_tx_details` and other functions, in particular, to `fetch_and_insert_txs`

Reply:
I have refactored almost the whole workflow, it's a little bit different than the past now